pyccel/parser/semantic.py

```diff
@@ -2659,7 +2659,7 @@ def _visit_FunctionDef(self, expr, **settings):
                             args_inout[i] = True
 
                         i_fa += 1
-                if (args_inout[i] is True and a.is_const is True) or (str(a) in all_assigned and a.is_const is True):
+                if (args_inout[i] or (str(a) in all_assigned)) and a.is_const:
```

Comment:
Just a minor detail...

As far as I know the Python interpreter processes a condition in the form `a and b` by first checking `a`, and then it checks `b` only if `a == True`. Hence you should put in `a` the easiest condition to evaluate, in this case `a.is_const`.

Reply:
when i write the condition like :
`if a.is_const and (args_inout[i] or (str(a) in all_assigned)):`
the `test_higherorder_functions.py` fails saying that :
` AttributeError: 'FunctionAddress' object has no attribute 'is_const'`
so i think i need also to check for` isinstance(a, Variable) == True`